benchmark/kernel/gcn-ogbn.py

```diff
@@ -0,0 +1,167 @@
+
+
+def main():
+    parser = argparse.ArgumentParser(description='OGBN-Products (GNN)')
+    parser.add_argument('--device', type=int, default=0)
+    parser.add_argument('--log_steps', type=int, default=1)
+    parser.add_argument('--use_sage', action='store_true')
+    parser.add_argument('--num_layers', type=int, default=3)
+    parser.add_argument('--hidden_channels', type=int, default=256)
+    parser.add_argument('--dropout', type=float, default=0.5)
+    parser.add_argument('--lr', type=float, default=0.01)
+    parser.add_argument('--epochs', type=int, default=300)
+    parser.add_argument('--runs', type=int, default=10)
+    parser.add_argument('--inference', action='store_true')
+    parser.add_argument('--profile', action='store_true')
+    args = parser.parse_args()
+    print(args)
+
+    device = f'cuda:{args.device}' if torch.cuda.is_available() else 'cpu'
+    device = torch.device(device)
+
+    dataset = PygNodePropPredDataset(name='ogbn-products', root='dataset/',
+                                     transform=T.ToSparseTensor())
+    data = dataset[0]
+
+    split_idx = dataset.get_idx_split()
+    train_idx = split_idx['train'].to(device)
+
+    model = Net(data.num_features, args.hidden_channels, dataset.num_classes,
+                args.num_layers, args.dropout, args.use_sage).to(device)
+
+    if not args.use_sage:
+        # Pre-compute GCN normalization.
+        adj_t = data.adj_t.set_diag()
+        deg = adj_t.sum(dim=1).to(torch.float)
+        deg_inv_sqrt = deg.pow(-0.5)
+        deg_inv_sqrt[deg_inv_sqrt == float('inf')] = 0
+        adj_t = deg_inv_sqrt.view(-1, 1) * adj_t * deg_inv_sqrt.view(1, -1)
+        data.adj_t = adj_t
+
+    data = data.to(device)
+
+    if not args.inference:
+        evaluator = Evaluator(name='ogbn-products')
+
+        for run in range(args.runs):
+            model.reset_parameters()
+            optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
+            for epoch in range(1, 1 + args.epochs):
+                loss = train(model, data, train_idx, optimizer)
+                result = test(model, data, split_idx, evaluator)
+
+                if epoch % args.log_steps == 0:
+                    train_acc, valid_acc, test_acc = result
+                    print(f'Run: {run + 1:02d}, '
+                          f'Epoch: {epoch:02d}, '
+                          f'Loss: {loss:.4f}, '
+                          f'Train: {100 * train_acc:.2f}%, '
+                          f'Valid: {100 * valid_acc:.2f}% '
+                          f'Test: {100 * test_acc:.2f}%')
+    else:
+        model.reset_parameters()
+        for epoch in range(1, 1 + args.epochs):
+            print("Epoch ", epoch)
+            if epoch == args.epochs:
+                e2e_inference = e2e_time()(inference)
+                e2e_inference(model, data)
+            else:
+                inference(model, data)
+
+        if args.profile:
+            profile_inference = torch_profile()(inference)
```

Comment:
Here again.

Reply:
Current `torch_profile` is both a decorator and context manager.